standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/ObjectStore.java

```diff
@@ -10330,6 +10330,108 @@ public boolean deletePartitionColumnStatistics(String catName, String dbName, St
+                + "  for which stats deletion is requested doesn't exist");
+      }
+      // Note: this does not verify ACID state; called internally when removing cols/etc.
+      //       Also called via an unused metastore API that checks for ACID tables.
+      MPartition mPartition = getMPartition(catName, dbName, tableName, partVals, mTable);
+      if (mPartition == null) {
+        throw new NoSuchObjectException("Partition " + partName
+                + " for which stats deletion is requested doesn't exist");
+      }
+      query = pm.newQuery(MPartitionColumnStatistics.class);
+      String filter;
+      String parameters;
+      if (colNames != null) {
+        filter = "partition.partitionName == t1 && partition.table.database.name == t2 && partition.table.tableName == t3 && "
+                + "t4.contains(colName) && partition.table.database.catalogName == t5" + (engine != null ? " && engine == t6" : "");
+        parameters = "java.lang.String t1, java.lang.String t2, java.lang.String t3, "
+                + "java.util.Collection t4, java.lang.String t5" + (engine != null ? ", java.lang.String t6" : "");
+      } else {
+        filter = "partition.partitionName == t1 && partition.table.database.name == t2 && partition.table.tableName == t3 && partition.table.database.catalogName == t4" + (engine != null ? " && engine == t5" : "");
+        parameters = "java.lang.String t1, java.lang.String t2, java.lang.String t3, java.lang.String t4" + (engine != null ? ", java.lang.String t5" : "");
+      }
+      query.setFilter(filter);
+      query.declareParameters(parameters);
+      if (colNames != null) {
+        query.setUnique(true);
+        for (String colName : colNames){
+          // trim the extra spaces, and change to lowercase
+          normalizeIdentifier(colName);
+        }
+        if (engine != null) {
+          mStatsObj =
+                  (MPartitionColumnStatistics) query.executeWithArray(partName.trim(),
+                          normalizeIdentifier(dbName),
+                          normalizeIdentifier(tableName),
+                          colNames,
+                          normalizeIdentifier(catName),
+                          engine);
+        } else {
+          mStatsObj =
+                  (MPartitionColumnStatistics) query.executeWithArray(partName.trim(),
+                          normalizeIdentifier(dbName),
+                          normalizeIdentifier(tableName),
+                          colNames,
+                          normalizeIdentifier(catName));
+        }
+        pm.retrieve(mStatsObj);
+        if (mStatsObj != null) {
+          pm.deletePersistent(mStatsObj);
+        } else {
+          throw new NoSuchObjectException("Column stats doesn't exist for table="
+                  + TableName.getQualified(catName, dbName, tableName) +
+                  " partition=" + partName + " col=" + String.join(", ", colNames));
+        }
+      } else {
```

Comment:
nit: can we merge the `if else`?

Reply:
One condition is (colNames != null) and the other one condition is (colNames == null). In this 2 conditions, the JDO queries are different. So it's clearer to not merge them for now.

Reply:
The difference between the two conditions is the input parameters of the `Query`, we can use a list to collect all parameters and then push it to `query.executeWithArray(...`, example:
```
List parameters = new ArrayList();
parameters.add(normalizeIdentifier(tableName));
parameters.add(normalizeIdentifier(dbName));
parameters.add(normalizeIdentifier(catName));
if (colNames != null && !colNames.isEmpty()) {
  parameters.add(colNames);
}
if (engine != null) {
  parameters.add(engine);
}
 mStatsObjColl =
                  (List<MTableColumnStatistics>) query.executeWithArray(parameters.toArray());
...
```

Reply:
I think we can keep the old way. The old way has a better readability.